### PR TITLE
Create a DSL health indicator as part of the health API

### PR DESCRIPTION
**Pull request by andreidan**
This adds a health indicator named `data_stream_lifecycle` that will detect
data stream backing indices that cannot make progress
(stagnating) due to repeatedly error-ing in their lifecycle
execution. 

The output of the indicator looks like this:
```
"data_stream_lifecycle" : {
    "status" : "yellow",
    "symptom" : "2 backing indices have repeatedly encountered errors whilst trying to advance in its lifecycle",
    "details" : {
      "stagnating_backing_indices_count" : 2,
      "stagnating_backing_indices" : [
        {
          "index_name" : ".ds-metrics-foo-2023.12.07-000002",
          "first_occurrence_timestamp" : 1701951305340,
          "retry_count" : 4
        },
        {
          "index_name" : ".ds-metrics-foo-2023.12.07-000001",
          "first_occurrence_timestamp" : 1701951305340,
          "retry_count" : 4
        }
      ],
      "total_backing_indices_in_error" : 2
    },
    "impacts" : [
      {
        "id" : "elasticsearch:health:dsl:impact:stagnating_backing_index",
        "severity" : 3,
        "description" : "Data streams backing indices cannot make progress in their lifecycle. The performance and stability of the indices and/or the cluster could be impacted.",
        "impact_areas" : [
          "deployment_management"
        ]
      }
    ],
    "diagnosis" : [
      {
        "id" : "elasticsearch:health:dsl:diagnosis:stagnating_dsl_backing_index",
        "cause" : "Some backing indices are repeatedly encountering errors in their lifecycle execution.",
        "action" : "Check the current status of the affected indices using the [GET /<affected_index_name>/_lifecycle/explain] API. Please replace the <affected_index_name> in the API with the actual index name (or the data stream name for a wider overview).",
        "help_url" : "https://ela.st/dsl-explain",
        "affected_resources" : {
          "indices" : [
            ".ds-metrics-foo-2023.12.07-000002",
            ".ds-metrics-foo-2023.12.07-000001"
          ]
        }
      }
    ]
  }
```

Documentation will follow in a subsequent PR.

### CODE REVIEW

File: modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/lifecycle/DataStreamLifecycleServiceIT.java [hunk 1]
```diff
@@ -569,6 +620,39 @@ public void testErrorRecordingOnRetention() throws Exception {
...
+            GetHealthAction.Response healthResponse = client().execute(GetHealthAction.INSTANCE, new GetHealthAction.Request(true, 1000))
+                .actionGet();
+            List<HealthIndicatorResult> masterIsStableIndicator = healthResponse.getIndicatorResults()
+                .stream()
+                .filter(indicator -> indicator.name().equals(StableMasterHealthIndicatorService.NAME))
+                .toList();
+            assertThat(masterIsStableIndicator.size(), is(1));
+            // if the cluster doesn't have a stable master we'll avoid asserting on the health report API as some indicators will not
+            // be computed
+            if (masterIsStableIndicator.get(0).status() == HealthStatus.GREEN) {
```
Comment (gmarouli) [inline modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/lifecycle/DataStreamLifecycleServiceIT.java:632]: Same here

File: modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/lifecycle/DataStreamLifecycleServiceIT.java [hunk 2]
```diff
@@ -598,6 +682,29 @@ public void testErrorRecordingOnRetention() throws Exception {
...
+
+            healthResponse = client().execute(GetHealthAction.INSTANCE, new GetHealthAction.Request(true, 1000)).actionGet();
+            masterIsStableIndicator = healthResponse.getIndicatorResults()
+                .stream()
+                .filter(indicator -> indicator.name().equals(StableMasterHealthIndicatorService.NAME))
+                .toList();
+            assertThat(masterIsStableIndicator.size(), is(1));
+            // if the cluster doesn't have a stable master we'll avoid asserting on the health report API as some indicators will not
+            // be computed
+            if (masterIsStableIndicator.get(0).status() == HealthStatus.GREEN) {
```
Comment (gmarouli) [inline modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/lifecycle/DataStreamLifecycleServiceIT.java:694]: Assuming you agree :p, same here.

**Review by gmarouli — CHANGES_REQUESTED** [review]
Thank you for working on this @andreidan , really looking forward to see it released! I have some small comments, other than them, it's in a really good shape!

**Comment by andreidan** [comment]
@elasticmachine run elasticsearch-ci/8.11.2 / bwc-snapshots

**Comment by andreidan** [comment]
@elasticmachine update branch

File: modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/lifecycle/DataStreamLifecycleServiceIT.java [hunk 3]
```diff
@@ -569,6 +620,39 @@ public void testErrorRecordingOnRetention() throws Exception {
...
+                .actionGet();
+            List<HealthIndicatorResult> masterIsStableIndicator = healthResponse.getIndicatorResults()
+                .stream()
+                .filter(indicator -> indicator.name().equals(StableMasterHealthIndicatorService.NAME))
+                .toList();
+            assertThat(masterIsStableIndicator.size(), is(1));
+            // if the cluster doesn't have a stable master we'll avoid asserting on the health report API as some indicators will not
+            // be computed
+            if (masterIsStableIndicator.get(0).status() == HealthStatus.GREEN) {
+                // the dsl indicator should turn the overall status yell;ow
```
[the remark this answers is not in the document]
Comment (gmarouli) [reply modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/lifecycle/DataStreamLifecycleServiceIT.java:616]: I think this one was missed :)

**Review by gmarouli — APPROVED** [review]
LGTM, just one leftover `yello;w` 🚀 Can't wait to see it live!

**Pull request merged**